src/ast/pass_manager.h

```diff
@@ -1,64 +1,371 @@
+      return type_id;
+    }
+  };
 
-// Context/config for passes
-//
-// Note: Most state should end up in the BPFtrace class instead of here
+  // State is an object which is the result of some specific pass.
+  //
+  // Passes are allowed to mutate the AST, and produce intermediate results
+  // that may be consumed by other passes. Each of these results must be a
+  // State, and will be immutable once it is produced.
+  class State {
+  public:
+    virtual ~State() = default;
+  };
 
-struct PassContext {
-public:
-  PassContext(BPFtrace &b, ASTContext &ast_ctx) : b(b), ast_ctx(ast_ctx) {};
-  BPFtrace &b;
-  ASTContext &ast_ctx;
+  // get fetches the given type; which must have been previously stored.
+  //
+  // N.B. In the future, these objects should be made immutable. Right now,
+  // this is preserved so that we can pass through the `BPFTrace` object which
+  // is consumed by *most* passes in a read-write fashion.
+  template <typename T>
+  T &get() const
+  {
+    int type_id = TypeId<T>::type_id();
+    if (state_.contains(type_id)) {
+      auto it = state_.find(type_id);
+      return *static_cast<T *>(it->second.get());
+    } else if (extern_state_.contains(type_id)) {
+      return static_cast<T &>(extern_state_.find(type_id)->second.get());
+    } else {
+      fail(type_id);
+    }
+  }
+
+private:
+  // for the failed lookup path, see above.
+  [[noreturn]] static void fail(int type_id);
+
+  // put emplaces the the value into the map; it must not exist already. In
+  // general this should not be called manually, it should only be called by
+  // individual passes during execution (the `run` function below).
+  template <typename T>
+  void put(T &&t)
+  {
+    int type_id = TypeId<T>::type_id();
+    state_.emplace(type_id, std::make_unique<T>(std::move(t)));
+  }
+
+  // put can also provide an external reference for the map; its lifecycle is
+  // managed outside the scope of the passes. While passes can produce results
+  // that are references, this should be generally discouraged (and this
+  // capability should eventually be removed). Like `put`, this should not be
+  // called directly and should only be called by `run`, below.
+  template <typename T>
+  void put(T &t)
+  {
+    int type_id = TypeId<T>::type_id();
+    extern_state_.emplace(type_id, t);
+  }
+
+  // Will register the given type id and the name, which is resolvable using
+  // lookup_type below. This is called exclusively by `type_id` above.
+  static int add_type(std::string &&type_name)
+  {
+    int type_id = next_type_id_++;
+    type_names_[type_id] = std::move(type_name);
+    return type_id;
+  }
+
+  // Returns the underlying type for this type_id. Note that the type must
+  // exist, or this will have undefined behavior. This is called exclusively
+  // by the pass manager, when statically checking pass types.
+  static const std::string &lookup_type(int type_id)
+  {
+    return type_names_[type_id];
```

Comment:
Does this really justify it's own function and comment?

Reply:
Dropped.